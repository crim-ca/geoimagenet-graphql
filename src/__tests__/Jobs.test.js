// @flow

import {
    BATCH_CREATION_JOBS_RESPONSE,
    SPECIFIC_BATCH_CREATION_JOB_FAILURE_LOGS_1,
    SPECIFIC_BATCH_CREATION_JOB_FAILURE_LOGS_2,
    SPECIFIC_BATCH_CREATION_JOB_FAILURE_LOGS_3,
    SPECIFIC_BATCH_CREATION_JOB_RESPONSE_1,
    SPECIFIC_BATCH_CREATION_JOB_RESPONSE_2,
    SPECIFIC_BATCH_CREATION_JOB_RESPONSE_3,
} from "./api_responses";

const {createTestClient} = require('apollo-server-testing');
const {ApolloServer} = require('apollo-server');
const {Jobs} = require('../datasources/Jobs');
const {typeDefs} = require('../schema');
const resolvers = require('../resolvers');
const gql = require('graphql-tag');

const GET_JOBS = gql`
    query jobs {
        jobs(process_id: "batch-creation") {
            id
            status
            status_message
            progress
        }
    }
`;

const jobs = new Jobs('', '');
const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({jobs}),
    context: () => ({user_id: 1}),
});
jobs.get = jest.fn((url) => {
    switch (url) {
        case 'processes/batch-creation/jobs':
            return BATCH_CREATION_JOBS_RESPONSE;
        case 'processes/batch-creation/jobs/job_1':
            return SPECIFIC_BATCH_CREATION_JOB_RESPONSE_1;
        case 'processes/batch-creation/jobs/job_2':
            return SPECIFIC_BATCH_CREATION_JOB_RESPONSE_2;
        case 'processes/batch-creation/jobs/job_3':
            return SPECIFIC_BATCH_CREATION_JOB_RESPONSE_3;
        case 'processes/batch-creation/jobs/job_1/logs':
            return SPECIFIC_BATCH_CREATION_JOB_FAILURE_LOGS_1;
        case 'processes/batch-creation/jobs/job_2/logs':
            return SPECIFIC_BATCH_CREATION_JOB_FAILURE_LOGS_2;
        case 'processes/batch-creation/jobs/job_3/logs':
            throw new Error('This job is supposed to be successful, this should not be called.');
    }
    throw new Error('Unmanaged api response.');
});
const {query} = createTestClient(server);

test('Jobs have relevant error messages inaccordance to their status', async () => {
    const res = await query({query: GET_JOBS});
    expect(res).toMatchSnapshot();
});
