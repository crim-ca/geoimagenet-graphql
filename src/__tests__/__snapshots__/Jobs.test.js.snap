// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jobs have relevant error messages in accordance to their status 1`] = `
Object {
  "data": Object {
    "jobs": Array [
      Object {
        "id": "job_1",
        "progress": 3,
        "status": "failed",
        "status_message": "Job <job_1> failed execution [builtins.ValueError: Missing input 'taxonomy_url' not matched from literal process input nor defaults.]..",
      },
      Object {
        "id": "job_2",
        "progress": 7,
        "status": "failed",
        "status_message": "Job <job_2> failed execution [builtins.AssertionError: invalid rasterfile paths specified: []]..",
      },
      Object {
        "id": "job_3",
        "progress": 7,
        "status": "succeeded",
        "status_message": "Job <job_3> done.",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Jobs have relevant error messages inaccordance to their status 1`] = `
Object {
  "data": Object {
    "jobs": Array [
      Object {
        "id": "job_1",
        "progress": 3,
        "status": "failed",
        "status_message": "Job <job_1> failed execution [builtins.ValueError: Missing input 'taxonomy_url' not matched from literal process input nor defaults.]..",
      },
      Object {
        "id": "job_2",
        "progress": 7,
        "status": "failed",
        "status_message": "Job <job_2> failed execution [builtins.AssertionError: invalid rasterfile paths specified: []]..",
      },
      Object {
        "id": "job_3",
        "progress": 7,
        "status": "succeeded",
        "status_message": "Job <job_3> done.",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`We create metrics from api response 1`] = `
Object {
  "data": Object {
    "public_benchmarks": Array [
      Object {
        "owner": "1",
        "result": Object {
          "metrics": Object {
            "top_1_accuracy": 0,
            "top_5_accuracy": 100,
          },
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
